### ISSUE TITLE
storage, computed: Fast create and drop materialized view leaves task running in computed

### ISSUE THREAD

**Issue by lluki** [MaterializeInc/materialize#15831]
### What version of Materialize are you using?

v0.30.0-dev

### How did you install Materialize?

Built from source

### What is the issue?

Executing a fast sequence of create + drop materialized view leaves a tokio task running in computed.

```
> CREATE TABLE t1 (f1 int);
> CREATE MATERIALIZED VIEW v1 AS SELECT MIN(f1) FROM t1; DROP MATERIALIZED VIEW v1;
```
will lead computed to forever try to fetch a snapshot that will never be available:

```
2022-11-01T14:43:16.646303Z  INFO mz_persist_client::internal::machine: snapshot sf2bd7d0d-4bff-469e-b956-6892afee97a3 as of Antichain { elements: [1667313447000] } not yet available for upper Antichain { elements: [1667313447000] } retrying in 16s
```
and this message will repeat every 16 seconds.

In this case `sf2...` is actually the shard of the materialized view, not of `t1`, when computed starts writing to the shard, it first fetches a snapshot of it, which fails in this case and enters an endless retry loop.

See also this [slack thread](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1667314342022869)

It might also be a bug in the envd<->computed interaction, could also be Team Compute.

**Comment by lluki** [comment]
Often the first couple of lines are:
```snapshot sf2.... as of Antichain { elements: [1667313447000] } not yet available for upper Antichain { elements: [0] } ...``` 
and then something (environmentd?) advances the upper to the exact `as of` that is requested, from there on all lines will be:
```snapshot sf2.... as of Antichain { elements: [1667313447000] } not yet available for upper Antichain { elements: [1667313447000] } ...```. Afaik it is correct that the snapshot can't be fetched in this situation, the upper has to be strictly greater.

But is it possible that whoever advanced the initial timestamp from 0 to `x` should have advanced it to `x+1` such that the initial snapshot read would always succeed?

**Comment by teskje** [comment]
To summarize our debugging, the issue is the following:

* The `persist_sink` internally creates a `persist_source`, to read back data from the target shard, to apply corrections as necessary.
* The `persist_source` reads a snapshot of the shard at the dataflow's `as_of` when it is initialized. For this it has to wait until the upper of the shard has advanced beyond `as_of`.
* Responsible for advancing the `as_of` of the target shard is the `persist_sink` itself, specifically the `append_batches` operator.
* If the dataflow is dropped very quickly after having been created, it is possible that this happens while the `persist_source` is still waiting for the upper to advance beyond the `as_of`. We then drop the dataflow token, which causes the other `persist_sink` operators (including `append_batches`) to terminate.
* Since `append_batches` has terminated, there is nothing that will advance the shards upper anymore, so the `persist_source` is stuck waiting for the snapshot and never terminates.
